### PR TITLE
[BugFix] fix resultset metadata of prepare protocol

### LINKED ISSUE #38672: Mysql Compatibility issues with prepared statements.
### Steps to reproduce the behavior (Required)
1. Start SR locally with docker, connect via MySQL
2. CREATE DATABASE test;
3. 
```
CREATE TABLE IF NOT EXISTS test.`test` (
    `workspace_id`  varchar(36) NOT NULL,
    `test_id`       varchar(36) NOT NULL
    )
    ENGINE=OLAP
    PRIMARY KEY(`workspace_id`,`test_id`)
    DISTRIBUTED BY HASH(`workspace_id`)
    PROPERTIES(
                  "enable_persistent_index" = "true",
                  "partition_live_number" = "1095"
              );
```
4. Sample golang program
```
func main() {
	ctx := context.Background()
	dbConnectionString := fmt.Sprintf("%s:%s@tcp(%s:%v)/test", "root", "", "127.0.0.1", 9030)
	db, err := sql.Open("mysql", dbConnectionString)
	if err != nil {
		log.WithError(err).Errorf(ctx, "Error creating DB")
	}
	rows, err := db.Query("select count(*) from test WHERE test_id=?", "1234")
	//rows, err := db.Query("select count(*) from test WHERE test_id='1234'") // This runs fine
	if err != nil {
		log.WithError(err).Errorf(ctx, "Error running query DB")
		return
	}
	defer rows.Close()
	for rows.Next() {
		var result int64
		rows.Scan(&result)
		fmt.Printf("Received %v as count\n", result)
	}
	if rows.Err() != nil {
		log.WithError(rows.Err()).Errorf(ctx, "Error getting rows %v", rows.Err())
	}
}
```

### Expected behavior (Required)
Sample program should return 0 from the select count(*) statement

### Real behavior (Required)
Errors out with `"error.kind":"Error 1064 (HY000)","error.message":"Error 1064 (HY000): Unsupported command(COM_STMT_PREPARE)"` 

### StarRocks version (Required)
 - 3.2.2

### PR DESCRIPTION
**Pull request by murphyatwork**
## Why I'm doing:
- According to [MySQL Spec](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_com_stmt_prepare.html#sect_protocol_com_stmt_prepare_response), the prepare statement needs to return the metadata of query output, which means the server needs to **analyze** the statement

## What I'm doing:
- Analyze the statement when prepare, to return the correct resultset metadata
- Treat the `Parameter` as string Type when inferencing type, since string is compatible with almost all types

## Out of scope
- Optimize the query performance of prepare-statement

Fixes #38672

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Review by Seaven — DISMISSED** [review]

**Review by packy92 — APPROVED** [review]

**Review by decster — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java [hunk 1]
```diff
@@ -1749,11 +1752,24 @@ private void sendStmtPrepareOK(PrepareStmt prepareStmt) throws IOException {
...
-            context.getState().setEof();
-        } else {
-            context.getMysqlChannel().flush();
-            context.getState().setStateType(MysqlStateType.NOOP);
+            MysqlEofPacket eofPacket = new MysqlEofPacket(context.getState());
+            eofPacket.writeTo(serializer);
+            context.getMysqlChannel().sendOnePacket(serializer.toByteBuffer());
         }
+
+        if (numColumns > 0) {
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/qe/StmtExecutor.java:1760]: why not make it `else if` clause to make people know that only 1 scenario will happen.

File: fe/fe-core/src/main/java/com/starrocks/analysis/Parameter.java [hunk 2]
```diff
@@ -55,10 +58,17 @@ public <R, C> R accept(AstVisitor<R, C> visitor, C context) {
...
     public Type getType() {
+        Type res;
         if (expr != null) {
-            return expr.getType();
+            res = expr.getType();
+        } else {
+            res = super.getType();
         }
-        return super.getType();
+        // use STRING as default type, since string is compatible with almost all types in the type inference
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/analysis/Parameter.java:67]: When will the type be Invalid?

Comment (imay) [reply fe/fe-core/src/main/java/com/starrocks/analysis/Parameter.java:67]: I'm a little worried that we can't get invalid type when this is really needed. For example if the SQL is not correct.

**Review by nshangyiming — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/analysis/Parameter.java (hunk 2 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/analysis/Parameter.java:67]: The type of `Parameter` is INVALID, bc we don't need its type when executing `PREPARE ` statement.

The type would be determined in the `EXECUTE` statement, in which the parameter is provided and we the inference the type bottomup

**Pull request merged**